File: python/zpy/src/zingg_py/zingg.py
"""End-to-end orchestrator — port of ``Zingg.scala``.

    cfg = ZinggConf(fields=[
        FieldDef("summary", MatchType.Text),
        FieldDef("description", MatchType.cve),
        FieldDef("priority", MatchType.Exact),
    ])
    z = Zingg(cfg)
    candidates = z.find_training_data(df)
    # label `candidates` -> `labeled` with z_label in {0.0, 1.0}
    model, tree = z.train(labeled)
    clusters = z.cluster(df, model, tree)
"""

from __future__ import annotations

from typing import Optional

import pandas as pd

from . import active_learning, blocking, features, model as model_mod, pairs as pair_mod
from .blocking import BlockingTree, Leaf, Node
from .clustering import connected_components
from .hashing import IdentityString, cold_start_hashes
from .schema import ZinggConf


class Zingg:
    def __init__(self, cfg: ZinggConf, link: bool = False):
        self.cfg = cfg
        self.link = link

    def find_training_data(
        self,
        df: pd.DataFrame,
        prior_model: Optional[model_mod.TrainedModel] = None,
        prior_tree: Optional[BlockingTree] = None,
        n: int = 30,
    ) -> pd.DataFrame:
        cfg = self.cfg
        with_ids = self._ensure_id(df)
        if prior_tree is not None:
            pairs = pair_mod.self_pairs(
                blocking.assign_blocks(with_ids, prior_tree), cfg, cross_source_only=self.link)
        else:
            pairs = pair_mod.cold_start_pairs(with_ids, cfg, cross_source_only=self.link)
        featured = features.add_features(pairs, cfg)
        if prior_model is not None:
            scored = model_mod.score(prior_model, featured, cfg)
            sampled = active_learning.query_uncertain(scored, cfg, n)
        else:
            sampled = active_learning.cold_start_sample(featured, n, cfg.sample_seed)
        return self._strip_pipeline_columns(sampled)

    def _strip_pipeline_columns(self, df: pd.DataFrame) -> pd.DataFrame:
        to_drop = ["z_poly_features", "z_prob", "rawPrediction",
                   self.cfg.score_col, self.cfg.prediction_col, "z_uncertainty"]
        return df.drop(columns=[c for c in to_drop if c in df.columns]).reset_index(drop=True)

    def train(self, labeled: pd.DataFrame):
        cfg = self.cfg
        pos = labeled[labeled[cfg.label_col] == 1.0]
        tree = self._seed_tree() if len(pos) < 2 else blocking.learn(self._rows_to_pair_maps(pos), cfg)
        model = model_mod.train(labeled, cfg)
        return model, tree

    def cluster(self, df: pd.DataFrame, model: model_mod.TrainedModel, tree: BlockingTree,
                block_strategy: str = "tree") -> pd.DataFrame:
        """Cluster ``df`` into entities.

        ``block_strategy`` controls candidate generation for the final clustering:
          - ``"tree"`` (default): the learned single-key blocking tree — bounded
            block size, high precision, the scalable choice.
          - ``"canopies"``: the multi-key cold-start canopies (as used at
            ``find_training_data`` time) — higher recall, so a cluster whose
            members the single key fails to co-locate is not split, at the cost
            of more candidate pairs (slower) and more false-edge / hub-merge risk.
        """
        cfg = self.cfg
        with_ids = self._ensure_id(df)
        if block_strategy == "canopies":
            pairs = pair_mod.cold_start_pairs(with_ids, cfg, cross_source_only=self.link)
        elif block_strategy == "tree":
            blocked = blocking.assign_blocks(with_ids, tree)
            pairs = pair_mod.self_pairs(blocked, cfg, cross_source_only=self.link)
        else:
            raise ValueError(f"block_strategy must be 'tree' or 'canopies', got {block_strategy!r}")
        featured = features.add_features(pairs, cfg)
        scored = model_mod.score(model, featured, cfg)
        if scored.empty:
            scored[cfg.prediction_col] = pd.Series([], dtype=float)
        else:
            scored[cfg.prediction_col] = (scored[cfg.score_col] >= cfg.threshold).astype(float)
        clustered = connected_components(scored, with_ids[cfg.id_col].tolist(), cfg)
        return with_ids.merge(clustered, on=cfg.id_col, how="left")

    def interactive_session(self, df: pd.DataFrame, prior_slope: float = 6.0, prior_precision: float = 1.0,
                            balance_prior: float = 1.0, seed: Optional[int] = 0):
        from .interactive_session import InteractiveSession
        cfg = self.cfg
        with_ids = self._ensure_id(df)
        pairs = pair_mod.cold_start_pairs(with_ids, cfg, cross_source_only=self.link)
        featured = features.add_features(pairs, cfg)
        return InteractiveSession(cfg, featured, prior_slope, prior_precision, balance_prior, seed)

    # ── internals ────────────────────────────────────────────────────────────

    def _ensure_id(self, df: pd.DataFrame) -> pd.DataFrame:
        if self.cfg.id_col in df.columns:
            return df.reset_index(drop=True)
        out = df.reset_index(drop=True).copy()
        out[self.cfg.id_col] = range(len(out))
        return out

    def _seed_tree(self) -> BlockingTree:
        for f in self.cfg.fields:
            if f.blockable:
                hs = [h for mt in f.match_types for h in cold_start_hashes(mt)]
                h = hs[0] if hs else IdentityString()
                return Node(f.name, h, {}, Leaf("seed"))
        return Leaf("root")

    def _rows_to_pair_maps(self, rows: pd.DataFrame):
        out = []
        for _, r in rows.iterrows():
            out.append({
                f.name: (r.get(f"{ZinggConf.LEFT_PREFIX}{f.name}"),
                         r.get(f"{ZinggConf.RIGHT_PREFIX}{f.name}"))
                for f in self.cfg.fields
            })
        return out

File: zscala/src/main/scala/zingg/zs/Zingg.scala
package zingg.zs

import org.apache.spark.ml.PipelineModel
import org.apache.spark.ml.feature.PolynomialExpansion
import org.apache.spark.sql.{DataFrame, Row}
import org.apache.spark.sql.functions.{col, lit, monotonically_increasing_id, when}
import org.apache.spark.storage.StorageLevel

/** End-to-end entry point.
  *
  * {{{
  *   val cfg = ZinggConf(fields = Seq(
  *     FieldDef("summary", MatchType.Text),
  *     FieldDef("description", MatchType.CveId),
  *     FieldDef("priority", MatchType.Exact)
  *   ))
  *   val z = new Zingg(cfg)
  *   val candidates = z.findTrainingData(df)
  *   // label `candidates` (human or LLM) -> `labeled` with `z_label` in {0.0, 1.0}
  *   val (model, tree) = z.train(labeled)
  *   val clusters = z.cluster(df, model, tree)
  * }}}
  */
class Zingg(cfg: ZinggConf, link: Boolean = false) extends Serializable {

  def findTrainingData(
      df: DataFrame,
      priorModel: Option[PipelineModel] = None,
      priorTree: Option[BlockingTree] = None,
      n: Int = 30
  ): DataFrame = {
    val withIds = ensureId(df)
    val tree = priorTree.getOrElse(seedTree)
    val blocked = BlockingTree.assignBlocks(withIds, tree)
    val pairs = PairBuilder.selfPairs(blocked, cfg, crossSourceOnly = link)
    val featured = Features.addFeatures(pairs, cfg)
    val sampled = priorModel match {
      case Some(m) =>
        val scored = Classifier.score(m, featured, cfg)
        ActiveLearning.queryUncertain(scored, cfg, n)
      case None =>
        cfg.sampleSeed match {
          case Some(s) => ActiveLearning.coldStartSample(featured, n, s)
          case None    => ActiveLearning.coldStartSample(featured, n)
        }
    }
    stripPipelineColumns(sampled)
  }

  /** Pipeline-output columns left over from a prior model's `transform`.
    * Stripped before returning candidates so callers can pass the result
    * straight back into `train` without colliding with the pipeline's own
    * output columns on the next fit.
    */
  private def stripPipelineColumns(df: DataFrame): DataFrame = {
    val toDrop = Seq("z_poly_features", "z_prob", "rawPrediction",
                     cfg.scoreCol, cfg.predictionCol, "z_uncertainty")
    toDrop.foldLeft(df)((d, c) => if (d.columns.contains(c)) d.drop(c) else d)
  }

  /** One-pair-at-a-time active-learning session. See
    * [[InteractiveSession]] for the labelling loop contract.
    *
    * The candidate pool is built once (block → pair → featurize) and
    * persisted; only the classifier fit + score is re-run between labels,
    * and that runs on `ec` in the background so the human's labelling time
    * absorbs the retrain latency. */
  def interactiveSession(
      df: DataFrame,
      priorSlope: Double = 6.0,
      priorPrecision: Double = 1.0
  ): InteractiveSession = {
    val withIds  = ensureId(df)
    val tree     = seedTree
    val blocked  = BlockingTree.assignBlocks(withIds, tree)
    val pairs    = PairBuilder.selfPairs(blocked, cfg, crossSourceOnly = link)
    val featured = Features.addFeatures(pairs, cfg).persist(StorageLevel.MEMORY_AND_DISK)
    val poly = new PolynomialExpansion()
      .setInputCol(Features.FeatureCol)
      .setOutputCol("z_poly_features")
      .setDegree(2)
    val pool = poly.transform(featured).collect()
    new InteractiveSession(cfg, featured, pool, priorSlope, priorPrecision)
  }

  def train(labeled: DataFrame): (PipelineModel, BlockingTree) = {
    val posRows = labeled.filter(col(cfg.labelCol) === lit(1.0)).collect()
    val tree =
      if (posRows.length < 2) seedTree
      else BlockingTreeLearner.learn(rowsToPairMaps(posRows), cfg)
    val model = Classifier.train(labeled, cfg)
    (model, tree)
  }

  def cluster(df: DataFrame, model: PipelineModel, tree: BlockingTree): DataFrame = {
    val withIds = ensureId(df)
    val blocked = BlockingTree.assignBlocks(withIds, tree)
    val pairs = PairBuilder.selfPairs(blocked, cfg, crossSourceOnly = link)
    val featured = Features.addFeatures(pairs, cfg)
    val scored = Classifier.score(model, featured, cfg)
    val matched = scored.withColumn(
      cfg.predictionCol,
      when(col(cfg.scoreCol) >= lit(cfg.threshold), lit(1.0)).otherwise(lit(0.0))
    )
    val ids = withIds.select(col(cfg.idCol))
    val clustered = Clustering.connectedComponents(matched, ids, cfg)
    withIds.join(clustered, Seq(cfg.idCol), "left")
  }

  private def ensureId(df: DataFrame): DataFrame =
    if (df.columns.contains(cfg.idCol)) df
    else df.withColumn(cfg.idCol, monotonically_increasing_id())

  private def seedTree: BlockingTree = {
    cfg.fields.find(_.blockable) match {
      case None => BlockingTree.Leaf("root")
      case Some(f) =>
        val h = Hash.registry.find(_.applicableTo.contains(f.matchType))
                    .getOrElse(Hash.IdentityString)
        BlockingTree.Node(f.name, h, Map.empty, BlockingTree.Leaf("seed"))
    }
  }

  private def rowsToPairMaps(rows: Array[Row]): Seq[Map[String, (Any, Any)]] =
    rows.toSeq.map { r =>
      cfg.fields.map { f =>
        val a = r.getAs[Any](s"${ZinggConf.LeftPrefix}${f.name}")
        val b = r.getAs[Any](s"${ZinggConf.RightPrefix}${f.name}")
        f.name -> (a, b)
      }.toMap
    }
}
